lib/CFRPageActions.jsm

```diff
@@ -0,0 +1,232 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const POPUP_NOTIFICATION_ID = "contextual-feature-recommendation";
+
+const DELAY_BEFORE_EXPAND_MS = 1000;
+const DURATION_OF_EXPAND_MS = 5000;
+
+/**
+ * A WeakMap from browsers to {host, recommendation} pairs. Recommendations are
+ * defined in the ExtensionDoorhanger.schema.json.
+ *
+ * A recommendation is specific to a browser and host and is active until the
+ * given browser is closed or the user navigates (within that browser) away from
+ * the host.
+ */
+const RecommendationMap = new WeakMap();
+
+/**
+ * A WeakMap from windows to their CFR PageAction.
+ */
+const PageActionMap = new WeakMap();
+
+/**
+ * We need one PageAction for each window
+ */
+class PageAction {
+  constructor(win) {
+    this.window = win;
+    this.urlbar = win.document.getElementById("urlbar");
+    this.container = win.document.getElementById("contextual-feature-recommendation");
+    this.button = win.document.getElementById("cfr-button");
+    this.label = win.document.getElementById("cfr-label");
+
+    this._popupStateChange = this._popupStateChange.bind(this);
+    this._collapse = this._collapse.bind(this);
+    this._handleClick = this._handleClick.bind(this);
+
+    // Saved timeout IDs for scheduled state changes, so they can be cancelled
+    this.stateTransitionTimeoutIDs = [];
+
+    this.container.onclick = this._handleClick;
+  }
+
+  show(notificationText, shouldExpand = false) {
+    this.container.hidden = false;
+
+    this.label.value = notificationText;
+
+    // We can safely get the width of the label even though the recommendation is
+    // collapsed; the label itself remains full width (with its overflow hidden)
+    const [{width}] = this.label.getClientRects();
+    this.urlbar.style.setProperty("--cfr-label-width", `${width}px`);
+
+    if (shouldExpand) {
+      this._clearScheduledStateChanges();
+
+      // After one second, expand
+      this._expand(DELAY_BEFORE_EXPAND_MS);
+
+      // Five seconds later, collapse again
+      this._collapse(DELAY_BEFORE_EXPAND_MS + DURATION_OF_EXPAND_MS);
+    }
+  }
+
+  hide() {
```

Comment:
It might be good to make this `async` for testing purposes, i.e. resolve `this_collapse` when the timeout has completed